examples/MagicWeather/app/src/main/java/com/revenuecat/sample/ui/paywall/PaywallAdapter.kt

```diff
@@ -35,19 +38,90 @@ class PaywallAdapter(
+
+                    val product = it.storeProduct
+                    val price = product.oneTimeProductPrice?.formattedPrice ?: "???"
+
+                    viewHolder.view.findViewById<TextView>(R.id.paywall_item_price).text = price
+                    viewHolder.view.findViewById<TextView>(R.id.paywall_item_best_offer).visibility = View.GONE
+                }
+                is PaywallItem.Option -> {
+                    viewHolder.view.findViewById<TextView>(R.id.paywall_item_options_title).visibility = View.GONE
+                    viewHolder.view.findViewById<ConstraintLayout>(R.id.purchasableLayout).visibility = View.VISIBLE
+
+                    val option = it.subscriptionOption
+                    val price = option.pricingPhases.map { phase ->
+                        "${phase.formattedPrice} for ${phase.billingPeriod}"
+                    }.joinToString(separator = " -> ")
+
+                    viewHolder.view.findViewById<TextView>(R.id.paywall_item_price).text = price
+                    viewHolder.view.findViewById<TextView>(R.id.paywall_item_best_offer).visibility =  if (it.defaultOffer) View.VISIBLE else View.GONE
+                }
+            }
         }
     }
 
-    override fun getItemCount() = offering?.availablePackages?.size ?: 0
+    override fun getItemCount() = theItems.size ?: 0
+
+    private val theItems: List<PaywallItem>
+        get() = offering?.availablePackages?.flatMap {
+
+            if (it.product.subscriptionOptions.isEmpty()) {
+                listOf(
+                    PaywallItem.Title(it.product.title),
+                    PaywallItem.Product(it.product)
+                )
+            } else {
+                val product = it.product
+                val title = product.defaultOption?.pricingPhases?.lastOrNull()?.billingPeriod?.let { period ->
+                    when (period) {
+                        "P1W" -> "Weekly"
+                        "P4W" -> "Every 4  Weeks"
+                        "P1M" -> "Monthly"
+                        "P3M" -> "Every 3 Months"
+                        "P6M" -> "Every 6 Months"
+                        "P1Y" -> "Yearly"
+                        else -> {
+                            null
+                        }
+                    }
+                } ?: product.title
+
+                mutableListOf(PaywallItem.Title(title)) + it.product.subscriptionOptions.map { option ->
+                    PaywallItem.Option(option, option == it.product.defaultOption)
+                }.sortedBy { option -> !option.defaultOffer }
+            }
+        } ?: emptyList()
 }
+
+sealed class PaywallItem {
+    data class Title(
+        val title: String
+    ) : PaywallItem()
+
+    data class Product(
+        val storeProduct: StoreProduct
+    ) : PaywallItem()
+
+    data class Option(
+        val subscriptionOption: SubscriptionOption,
+        val defaultOffer: Boolean
+    ) : PaywallItem()
+}
```

Comment:
This is my attempt to show different types of items that display differently inside of a list adapter 😇 I didn't want to but the subscription titles ("Monthly", "Yearly", "Lifetime", etc) needed to be separated from the clickable buttons for `StoreProduct` (one time purchases) and `SubscriptionOption` (subs)

Reply:
i think it makes sense!